File: aoc2208.cc
#include "iostream"
#include "vector"
#include "sstream"

using	namespace std;

int	main()
{
	vector<vector<int>>	a;
	string			s;

	while (cin >> s)
	{
		stringstream	ss(s);
		vector<int>	t;
		char		c;

		while (ss >> c)
			t.push_back(c - '0');
		a.push_back(t);
	}

	int	R = (int) a.size(), C = (int) a[0].size();
	int	i, j, r, c;
	int	peri = R * 2 + C * 2 - 4;
	int	res = peri;
	
	// part 2
	long long	res2 = 0, scen;
	int		u, d, l, rr;
	
	r = 0;
	while (++r < R - 1)
	{
		c = 0;	
		while (++c < C - 1)
		{
			int	n = a[r][c];
			bool	ok;
			
			i = -1; // up
			ok = true;
			while (++i < r)
			{
				if (a[i][c] >= n)
					ok = false;
			}
			if (ok)
			{
				res++;
				continue ;
			}
			i = r; // down
			ok = true;
			while (++i < R)
			{
				if (a[i][c] >= n)
					ok = false;
			}
			if (ok)
			{
				res++;
				continue ;
			}
			ok = true;
			i = -1; // left
			while (++i < c)
			{
				if (a[r][i] >= n)
					ok = false;
			}
			if (ok)
			{
				res++;
				continue ;
			}
			ok = true;
			i = c; // right
			while (++i < C)
			{
				if (a[r][i] >= n)
					ok = false;
			}
			if (ok)
			{
				res++;
				continue ;
			}
		}

		// part 2
		
		c = 0;
		while (++c < C - 1)
		{
			int	n = a[r][c];
			
			u = 0;
			i = r;
			while (--i > -1)
			{
				u++;
				if (a[i][c]>=n)
					break;
			}
			d = 0;
			i = r;
			while (++i < R)
			{
				d++;
				if (a[i][c]>=n)
					break;
			}
			l = 0;
			i = c;
			while (--i > -1)
			{
				l++;
				if (a[r][i]>=n)
					break;
			}
			rr = 0;
			i = c;
			while (++i < C)
			{
				rr++;
				if (a[r][i]>=n)
					break ;
			}
			scen = u * d * l * rr;
			res2 = res2 > scen ? res2 : scen;
		}
	}

	cout << "Star 1: " << res << endl;
	cout << "Star 2: " << res2 << endl;
} 

File: aoc2208.py
a = []
with open('2208.0') as file:
# with open('test') as file:
    for line in file:
        temp = []
        for c in line.strip():
            temp.append(int(c))
        a.append(temp)
R, C = len(a), len(a[0])
peri = R * 2 + C * 2 - 4
res = peri
res2 = 0
for r in range(1, R - 1):
    for c in range(1, C - 1):
        n = a[r][c]
        # up
        ok = True
        for i in range(r):
            if a[i][c] >= n:
                ok = False
        if ok:
            res += 1
            continue
        # down
        ok = True
        for i in range(r + 1, R):
            if a[i][c] >= n:
                ok = False
        if ok:
            res += 1
            continue
        # left
        ok = True
        for i in range(c):
            if a[r][i] >= n:
                ok = False
        if ok:
            res += 1
            continue
        # right
        ok = True
        for i in range(c + 1, C):
            if a[r][i] >= n:
                ok = False
        if ok:
            res += 1
            continue
    
    # part 2
    
    for c in range(1, C - 1):
        n = a[r][c]
        u = 0
        for i in range(r - 1, -1, -1):
            u += 1
            if a[i][c] >= n:
                break
        d = 0
        for i in range(r + 1, R):
            d += 1
            if a[i][c] >= n:
                break
        l = 0
        for i in range(c - 1, -1, -1):
            l += 1
            if a[r][i] >= n:
                break
        rr = 0
        for i in range(c + 1, C):
            rr += 1
            if a[r][i] >= n:
                break
        temp = u * d * l * rr
        res2 = res2 if res2 > temp else temp
print('Star 1:', res)
print('Star 2:', res2)
